src/main/java/com/starrocks/connector/flink/table/sink/StarRocksDynamicSinkFunctionV2.java

```diff
@@ -195,36 +181,65 @@ public void snapshotState(FunctionSnapshotContext functionSnapshotContext) throw
         StreamLoadSnapshot snapshot = sinkManager.snapshot();
 
         if (sinkManager.prepare(snapshot)) {
-            snapshotMap.put(functionSnapshotContext.getCheckpointId(), snapshot);
+            snapshotMap.put(functionSnapshotContext.getCheckpointId(), Collections.singletonList(snapshot));
 
             snapshotStates.clear();
-            snapshotStates.add(Tuple2.of(functionSnapshotContext.getCheckpointId(), snapshot));
+            snapshotStates.add(StarrocksSnapshotState.of(snapshotMap));
         } else {
             throw new RuntimeException("Snapshot state failed by prepare");
         }
     }
 
     @Override
     public void initializeState(FunctionInitializationContext functionInitializationContext) throws Exception {
-        ListStateDescriptor<Tuple2<Long, StreamLoadSnapshot>> descriptor =
+        ListStateDescriptor<StarrocksSnapshotState> descriptor =
                 new ListStateDescriptor<>(
                         "starrocks-sink-transaction",
-                        TypeInformation.of(new TypeHint<Tuple2<Long, StreamLoadSnapshot>>() {})
+                        TypeInformation.of(new TypeHint<StarrocksSnapshotState>() {})
                 );
         snapshotStates = functionInitializationContext.getOperatorStateStore().getListState(descriptor);
+
+        if (functionInitializationContext.isRestored()) {
+            for (StarrocksSnapshotState state : snapshotStates.get()) {
+                for (Map.Entry<Long, List<StreamLoadSnapshot>> entry : state.getData().entrySet()) {
+                    snapshotMap.compute(entry.getKey(), (k, v) -> {
+                        if (v == null) {
+                            return new ArrayList<>(entry.getValue());
+                        }
+                        v.addAll(entry.getValue());
+                        return v;
+                    });
+                }
+            }
+        }
     }
 
     @Override
     public void notifyCheckpointComplete(long checkpointId) throws Exception {
-        StreamLoadSnapshot snapshot = snapshotMap.get(checkpointId);
-        if (snapshot == null) {
-            log.warn("Unknown checkpoint id : {}", checkpointId);
-            return;
+
+        boolean succeed = true;
+
+        List<Long> commitCheckpointIds = snapshotMap.keySet().stream()
```

Comment:
Could the commitCheckpointIds be a empty list?

Reply:
empty is possible